File: internal/gateway/connector/platform_support.go
// Copyright 2026 Cisco Systems, Inc. and its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package connector

import (
	"fmt"
	"runtime"
)

// proxyConnectors are the chat/LLM-proxy connectors that interpose on model
// traffic through DefenseClaw's local guardrail proxy. They are unsupported on
// Windows, where DefenseClaw runs hook-only: the native Go hook entrypoint
// replaces the Bash hook chain, and there is no Windows guardrail-proxy
// lifecycle to host these connectors.
//
// This is the Go single source of truth for OS support. Keep it in sync with
// the Python cli/defenseclaw/platform_support.py WINDOWS_UNSUPPORTED_CONNECTORS
// set — a parity test pins the two lists together.
var proxyConnectors = map[string]struct{}{
	"openclaw":  {},
	"zeptoclaw": {},
}

// IsProxyConnector reports whether name is a proxy/chat connector (as opposed
// to a hook-based connector).
func IsProxyConnector(name string) bool {
	_, ok := proxyConnectors[name]
	return ok
}

// connectorSupportedOnOS reports whether a connector can be listed or set up on
// the given GOOS. Every hook-based connector is supported on every OS; the
// proxy connectors are unsupported on Windows.
func connectorSupportedOnOS(name, goos string) bool {
	if goos == "windows" && IsProxyConnector(name) {
		return false
	}
	return true
}

// ConnectorSupportedOnHostOS reports support on the current host OS.
func ConnectorSupportedOnHostOS(name string) bool {
	return connectorSupportedOnOS(name, runtime.GOOS)
}

// errConnectorUnsupportedOnOS is the clear, behavior-first error proxy-connector
// Setup returns when invoked on an OS that cannot host the guardrail proxy.
func errConnectorUnsupportedOnOS(name, goos string) error {
	return fmt.Errorf("connector %q is not supported on %s: DefenseClaw on %s is hook-only and proxy connectors require the guardrail proxy", name, goos, goos)
}

File: cli/defenseclaw/platform_support.py
# Copyright 2026 Cisco Systems, Inc. and its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# SPDX-License-Identifier: Apache-2.0

"""Per-OS connector support — the Python single source of truth.

DefenseClaw runs hook-only on Windows: agents invoke the native Go hook
entrypoint (``defenseclaw hook``) directly, and there is no Windows
guardrail-proxy lifecycle. The proxy/chat connectors (``openclaw`` and
``zeptoclaw``) therefore cannot run on Windows, so the TUI/CLI must not
offer or accept them there.

This module mirrors the Go ``connector.proxyConnectors`` set in
``internal/gateway/connector/platform_support.go``; a parity test pins the
two together so the lists cannot drift.

Filtering is a pure no-op on non-Windows hosts, so callers can apply it
unconditionally at presentation/validation points without changing
behavior on macOS or Linux.
"""

from __future__ import annotations

import sys
from collections.abc import Iterable

# Proxy/chat connectors that require the local guardrail proxy. These are
# the only connectors unsupported on Windows. Keep in sync with the Go
# ``proxyConnectors`` map.
WINDOWS_UNSUPPORTED_CONNECTORS: frozenset[str] = frozenset({"openclaw", "zeptoclaw"})


def host_os() -> str:
    """Return a Go-``GOOS``-style token for the current host.

    Normalizes ``sys.platform`` to ``"windows"`` / ``"darwin"`` / ``"linux"``
    so the same string compares cleanly against the Go side and against the
    ``os_name`` arguments below.
    """
    plat = sys.platform
    if plat.startswith("win"):
        return "windows"
    if plat == "darwin":
        return "darwin"
    if plat.startswith("linux"):
        return "linux"
    return plat


def is_proxy_connector(name: str) -> bool:
    """Report whether *name* is a proxy/chat connector."""
    return name in WINDOWS_UNSUPPORTED_CONNECTORS


def connector_supported_on_os(name: str, os_name: str | None = None) -> bool:
    """Report whether connector *name* can be offered/used on *os_name*.

    Every hook-based connector is supported on every OS; the proxy
    connectors are unsupported on Windows. *os_name* defaults to the host
    OS but is injectable so tests can assert behavior for any platform.
    """
    if os_name is None:
        os_name = host_os()
    if os_name == "windows" and name in WINDOWS_UNSUPPORTED_CONNECTORS:
        return False
    return True


def supported_connectors(
    names: Iterable[str], os_name: str | None = None
) -> list[str]:
    """Filter *names* down to those supported on *os_name*, preserving order."""
    if os_name is None:
        os_name = host_os()
    return [n for n in names if connector_supported_on_os(n, os_name)]
